fastestimator/architecture/unet.py

```diff
@@ -12,66 +12,151 @@
+    assert dropout is None or 0 <= dropout <= 1, \
+            "Invalid value for dropout parameter (None or 0 to 1 only)"  
+
+    assert bn in [None, "before", "after"], "Invalid bn parameter value"
+    
+    assert len(nchannels) >= 2, "At least 2 channels necessary for UNet"
+
+    conv_config = {'activation': activation, 'padding': 'same', 'kernel_initializer': 'he_normal'}
+    
     inputs = Input(input_size)
-    conv1 = Conv2D(64, 3, **conv_config)(inputs)
-    conv1 = Conv2D(64, 3, **conv_config)(conv1)
-    pool1 = MaxPooling2D(pool_size=(2, 2))(conv1)
-
-    conv2 = Conv2D(128, 3, **conv_config)(pool1)
-    conv2 = Conv2D(128, 3, **conv_config)(conv2)
-    pool2 = MaxPooling2D(pool_size=(2, 2))(conv2)
-
-    conv3 = Conv2D(256, 3, **conv_config)(pool2)
-    conv3 = Conv2D(256, 3, **conv_config)(conv3)
-    pool3 = MaxPooling2D(pool_size=(2, 2))(conv3)
-
-    conv4 = Conv2D(512, 3, **conv_config)(pool3)
-    conv4 = Conv2D(512, 3, **conv_config)(conv4)
-    drop4 = Dropout(0.5)(conv4)
-    pool4 = MaxPooling2D(pool_size=(2, 2))(drop4)
-
-    conv5 = Conv2D(1024, 3, **conv_config)(pool4)
-    conv5 = Conv2D(1024, 3, **conv_config)(conv5)
-    drop5 = Dropout(0.5)(conv5)
-
-    up6 = Conv2D(512, 2, **conv_config)(UpSampling2D(size=(2, 2))(drop5))
-    merge6 = concatenate([drop4, up6], axis=3)
-    conv6 = Conv2D(512, 3, **conv_config)(merge6)
-    conv6 = Conv2D(512, 3, **conv_config)(conv6)
-
-    up7 = Conv2D(256, 2, **conv_config)(UpSampling2D(size=(2, 2))(conv6))
-    merge7 = concatenate([conv3, up7], axis=3)
-    conv7 = Conv2D(256, 3, **conv_config)(merge7)
-    conv7 = Conv2D(256, 3, **conv_config)(conv7)
-
-    up8 = Conv2D(128, 2, **conv_config)(UpSampling2D(size=(2, 2))(conv7))
-    merge8 = concatenate([conv2, up8], axis=3)
-    conv8 = Conv2D(128, 3, **conv_config)(merge8)
-    conv8 = Conv2D(128, 3, **conv_config)(conv8)
-
-    up9 = Conv2D(64, 2, **conv_config)(UpSampling2D(size=(2, 2))(conv8))
-    merge9 = concatenate([conv1, up9], axis=3)
-    conv9 = Conv2D(64, 3, **conv_config)(merge9)
-    conv9 = Conv2D(64, 3, **conv_config)(conv9)
-    conv9 = Conv2D(2, 3, **conv_config)(conv9)
-    conv10 = Conv2D(1, 1, activation='sigmoid')(conv9)
-    model = Model(inputs=inputs, outputs=conv10)
+    inp = inputs
+
+    levels = []
+
+    # Contracting blocks
+    for idx, nc in enumerate(nchannels[:-1]):
+        
+        if idx == len(nchannels)-2:
+            d=dropout
+        else:
+            d=None
+
+        C, C_pooled = conv_block(inputs, nc, 3, conv_config, pooling=2, dropout=d, bn=bn)
+        levels.append((C, C_pooled))
+        inputs = C_pooled
+
+    # Expanding blocks
+    inp1, inp2 = levels[-1][1], levels[-1][0]
+    for idx, nc in enumerate(reversed(nchannels[1:])):
+        if idx == 0:
+            d = dropout
+        else:
+            d = None
+        
+        D = up_concat(inp1, inp2, 2, (nchannels[-1-idx], nchannels[-2-idx]), 3, conv_config, d, bn=bn)
```

Comment:
does dropout go in both the up and down? I only see it in one place in the old file